### PR TITLE
Datepicker

### PR DESCRIPTION
**Pull request by quirk0o**
@adamzima @s1mplex @lkudyba @jtkpiotr

### CODE REVIEW

File: client/.eslintrc.yml [hunk 1]
```diff
@@ -15,6 +15,7 @@ env:
   browser: true
   node: true
   mocha: true
+  jquery: true
```
Comment (adamzima) [inline client/.eslintrc.yml:18]: :(

File: client/app/components/shared/time/datepicker/TimeInput.jsx [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.value.getTime() !== this.props.value.getTime()) {
+      this._updatePicker(nextProps.value);
+    }
+  }
+
+  render() {
+    return (
+      <input type="text"
+             className={`time form-control ${this.props.className}`}
```
Comment (madejejej) [inline client/app/components/shared/time/datepicker/TimeInput.jsx:36]: Do you need `{}` brackets here?

File: client/app/helpers/DateHelper.js [hunk 3]
```diff
@@ -87,6 +88,16 @@ export function sameDay(date, other) {
   return moment(date).day() === moment(other).day();
 }
 
+
+export function roundedTime(date, granularity) {
+  const diff = date.unix() % granularity;
+  if (diff > 60) {
```
Comment (madejejej) [inline client/app/helpers/DateHelper.js:93]: I don't get the magic number 60 here. There's a test case called "with time almost equal to a multiple of granularity" that covers this. Why are you not using a value like `granularity / 2` instead? If this is intended then I think extracting 60 to a well-named constant would clarify a lot.

**Review by madejejej — APPROVED** [review]

File: client/app/components/shared/time/datepicker/TimeInput.jsx [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      minTime: this.props.minTime
+    }));
+  }
+
+  _bindChangeListener() {
+    this.input.on('changeTime', () => {
+      if (this.props.onChange) this.props.onChange(this.input.timepicker('getTime'));
+    });
+    this.input.on('timeFormatError', () => {
+      this.props.onChange();
```
Comment (jtkpiotr) [inline client/app/components/shared/time/datepicker/TimeInput.jsx]: It is not so obvious what it does. You could replace it with some other method, like: setDefaultTime or something

File: app/models/google_calendar/event_wrapper/builder.rb [hunk 5]
```diff
@@ -31,11 +31,7 @@ def build_rounded_event_wrapper
                     :end_time, :attendees, :conference_room, :user_email, :creator
 
       def build_date_time(date_time)
-        Google::Apis::CalendarV3::EventDateTime.new(date_time: format_date_time(date_time)) if date_time
-      end
-
-      def format_date_time(time)
-        DateTime.parse("#{time} #{time_zone}")
+        Google::Apis::CalendarV3::EventDateTime.new(date_time: date_time)
```
Comment (adamzima) [inline app/models/google_calendar/event_wrapper/builder.rb:34]: You'll have conflict with recurring PR

File: client/app/components/calendar/modal/layout/ModalBody.jsx [hunk 6]
```diff
@@ -15,12 +15,9 @@ export default class ModalBody extends React.Component {
     availableLocations: array,
     unavailableLocations: array,
     selectedLocation: number,
-    startTime: string.isRequired,
-    endTime: string.isRequired,
-    dateFormat: string.isRequired,
     errors: object,
     showErrorMessage: bool,
-    onError: func.isRequired
+    onGuestsError: func.isRequired
```
Comment (adamzima) [inline client/app/components/calendar/modal/layout/ModalBody.jsx]: Why you've changed the name?

File: client/app/components/calendar/modal/layout/body/FormDateField.jsx [hunk 7]
```diff
@@ -1,33 +1,27 @@
...
     </ControlLabel>
-    <DateTimeField
-      dateTime={value}
-      format={dateFormat}
-      inputFormat={dateFormat}
-      onChange={onChange} />
-    <If condition={errors !== undefined}>
+    <DateRangePicker onChange={onChange} errors={errors} />
+    <If condition={errors != null}>
       <div className="text-danger">{errors[0]}</div>
```
Comment (adamzima) [inline client/app/components/calendar/modal/layout/body/FormDateField.jsx]: Do you want to display empty div anyway?

File: client/app/components/calendar/modal/layout/body/FormTextField.jsx [hunk 8]
```diff
@@ -2,23 +2,24 @@ import React from 'react';
...
 
-const { string, func } = React.PropTypes;
+const { string, func, bool } = React.PropTypes;
 
-const FormTextField = (props) => (
+const FormTextField = ({ name, onChange, required }) => (
   <FormGroup>
     <ControlLabel>
-      {_.capitalize(props.name)}:
+      {_.capitalize(name)}:{required ? '*' : ''}
```
Comment (adamzima) [inline client/app/components/calendar/modal/layout/body/FormTextField.jsx]: `{required ? '*' : ''}` everywhere. Any idea how to make it more DRY?

File: client/app/components/shared/time/datepicker/DateInput.jsx [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.value.getTime() !== this.props.value.getTime()) {
+      this._updatePicker(nextProps.value);
+    }
+  }
+
+  render() {
+    return (<input type="text"
+                  className={`date form-control ${this.props.className}`}
```
Comment (adamzima) [inline client/app/components/shared/time/datepicker/DateInput.jsx]: weird formatting

File: client/app/components/shared/time/datepicker/DateRangePicker.jsx [hunk 10]
```diff
@@ -0,0 +1,84 @@
+import moment from 'moment';
+import React from 'react';
+import { roundedTime, addDateAndTime } from 'helpers/DateHelper';
+import DateInput from './DateInput';
+import TimeInput from './TimeInput';
+
+import './datepicker.scss';
+
+const INITIAL_TIME = roundedTime(moment(), 60 * 30);
```
Comment (adamzima) [inline client/app/components/shared/time/datepicker/DateRangePicker.jsx]: magic number `60 * 30`

File: client/app/components/shared/time/datepicker/TimeInput.jsx [hunk 11]
```diff
@@ -0,0 +1,59 @@
+import assign from 'lodash/assign';
+import React from 'react';
+
+const { string, func, instanceOf, bool } = React.PropTypes;
+
+const PICKER_OPTIONS = {
+  timeFormat: 'G:i'
```
Comment (adamzima) [inline client/app/components/shared/time/datepicker/TimeInput.jsx:7]: <3

**Review by adamzima — CHANGES_REQUESTED** [review]

File: client/app/components/shared/time/datepicker/TimeInput.jsx (hunk 2 above)
Comment (quirk0o) [reply client/app/components/shared/time/datepicker/TimeInput.jsx:36]: Yes

File: client/app/components/shared/time/datepicker/TimeInput.jsx (hunk 4 above)
Comment (quirk0o) [reply client/app/components/shared/time/datepicker/TimeInput.jsx]: _[Henryk]_ How about no?

File: client/app/components/calendar/modal/layout/body/FormTextField.jsx (hunk 8 above)
Comment (quirk0o) [reply client/app/components/calendar/modal/layout/body/FormTextField.jsx]: Done

File: client/app/components/shared/time/datepicker/DateInput.jsx (hunk 9 above)
Comment (quirk0o) [reply client/app/components/shared/time/datepicker/DateInput.jsx]: Done

File: client/app/components/shared/time/datepicker/TimeInput.jsx (hunk 4 above)
Comment (quirk0o) [reply client/app/components/shared/time/datepicker/TimeInput.jsx]: Done

File: app/models/google_calendar/event_wrapper/builder.rb (hunk 5 above)
Comment (quirk0o) [reply app/models/google_calendar/event_wrapper/builder.rb:34]: What are we supposed to do about it?

File: client/app/components/calendar/modal/layout/ModalBody.jsx (hunk 6 above)
Comment (quirk0o) [reply client/app/components/calendar/modal/layout/ModalBody.jsx]: Check again - we now have two error callbacks for date and location

File: client/app/components/calendar/modal/layout/body/FormDateField.jsx (hunk 7 above)
Comment (quirk0o) [reply client/app/components/calendar/modal/layout/body/FormDateField.jsx]: Done

File: client/app/components/shared/time/datepicker/DateRangePicker.jsx (hunk 10 above)
Comment (quirk0o) [reply client/app/components/shared/time/datepicker/DateRangePicker.jsx]: Done

**Review by adamzima — APPROVED** [review]

**Pull request merged**